### PR TITLE
optimizer: hoist type knowledge into transform::normalize_lets

### LINKED ISSUE #17762: [SQLsmith] mz_transform: Optimizer::optimize(): internal transform error: scalar types do not match: [ColumnType { scalar_type: UInt64, nullable: false }, ...]
### What version of Materialize are you using?

v0.45.0-dev (d6371fdc5)

### How did you install Materialize?

Built from source

### What is the issue?

```
materialize=> select from mz_internal.mz_scheduling_elapsed_internal join mz_indexes on(NULL)
    join(select worker_id a,
         cast(nullif(records, cast(null as numeric)) as numeric)
             from mz_internal.mz_records_per_dataflow) on(select a is NULL)
        where(select a) is NULL;
ERROR:  internal transform error: scalar types do not match: [ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: Oid, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: Some(NumericMaxScale(0)) }, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: Some(NumericMaxScale(0)) }, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Bool, nullable: false }] v [ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: Oid, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: Some(NumericMaxScale(0)) }, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: None }, nullable: true }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Bool, nullable: false }]
```

### Relevant log output

```source
environmentd: 2023-02-21T14:36:40.011348Z ERROR mz_transform: Optimizer::optimize(): internal transform error: scalar types do not match: [ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: Oid, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: Some(NumericMaxScale(0)) }, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: Some(NumericMaxScale(0)) }, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Bool, nullable: false }] v [ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: Oid, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: String, nullable: false }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: Some(NumericMaxScale(0)) }, nullable: false }, ColumnType { scalar_type: Numeric { max_scale: None }, nullable: true }, ColumnType { scalar_type: UInt64, nullable: false }, ColumnType { scalar_type: Bool, nullable: false }]
```

### PR DESCRIPTION
**Pull request by sploiselle**
Our type system has an unexpected corner in that the Eq implementation is derived but has no type knowledge, which necessitates using something other than `==` when determining type interoperability, i.e. a function named `base_eq`.

This patch fixes an instance of this in the optimizer, but probably warrants a more thorough investigation into where else it's used.

### Motivation

This PR fixes a recognized bug. Fixes #17762

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Comment by aalexandrov** [comment]
@sploiselle I've tried adding the derived `eq` to the list of `disallowed-methods` in `clippy.yoml`, sadly without `cargo clippy` complaining about the `==` comparison that you are fixing. It seems that we have to wait until #8581 before we can merge this.

```yaml
disallowed-methods = [
	// ...
    { path = "mz_repr::ScalarType::eq", reason = "use `ScalarType::base_eq` instead" },
]
```

**Review by aalexandrov — APPROVED** [review]

**Comment by frankmcsherry** [comment]
Thanks @sploiselle!

**Pull request merged**
